### ISSUE TITLE
Add possibility to define custom fasthttp server MaxRequestBodySize parameter to handle uploading of big files.

### ISSUE THREAD

**Issue by apuzyrevsky** [dapr/dapr#2659]
## In what area(s)?
> /area runtime

## Describe the feature

During the writing and deployment of the service which uploads files (up to 2GB), I've encountered an issue that files bigger than 4 megabytes are not uploading. I've gone on an investigation where exactly the issue lies and eventually find out that dapr using for HTTP server fasthttp library, which by default got a 4 MB limit on max body size and in dapr code it seems that default fasthttp server is defined. 
So, my question is - is it possible to add the possibility of providing MaxRequestBodySize parameter of fasthttp via annotation for example. I think I can lead that effort and provide a pull request if you guys would be okay with that kind of feature.

## Release Note
RELEASE NOTE:
Add the possibility of providing MaxRequestBodySize parameter of fasthttp server via annotation to handle uploading of big files.

**Comment by yaron2** [comment]
Thanks for raising this, and we would gladly appreciate a PR.

This would need to propagate to Dapr in the form of command line flag, I suggest using `-dapr-http-max-request-size`.

The CLI flags are located here: https://github.com/dapr/dapr/blob/master/pkg/runtime/cli.go

**Comment by apuzyrevsky** [comment]
@yaron2 Thank you for the fast reply. I've figured out about the CLI flag, although it's still not clear to me how can I pass it via annotation in yaml. Is that located here or I looking in not the right way? https://github.com/dapr/dapr/blob/master/pkg/injector/pod_patch.go

**Comment by yaron2** [comment]
> @yaron2 Thank you for the fast reply. I've figured out about the CLI flag, although it's still not clear to me how can I pass it via annotation in yaml. Is that located here or I looking in not the right way? https://github.com/dapr/dapr/blob/master/pkg/injector/pod_patch.go

Yes, the args are set here: https://github.com/dapr/dapr/blob/4678e477562e7e35a1d6ba04a9b17b6c9c00a025/pkg/injector/pod_patch.go#L524

**Comment by apuzyrevsky** [comment]
Hi @yaron2 
I've made appropriate changes and during testing encountered on the two new issues.
1. When I trying pass to my upload service via darp file more than 4 MB and less than 16 MB (I'll explain it in the next item) receiving the following error:
`{
    "errorCode": "ERR_DIRECT_INVOKE",
    "message": "rpc error: code = ResourceExhausted desc = grpc: received message larger than max (10730238 vs. 4194304)"
}` 
Do you have any ideas on how can I fix that as well? 
2. When I pass to my upload service via darp file more than 16 MB (seems that it is a default value for the maximum file size in memory https://github.com/valyala/fasthttp/blob/70e00dc4f3b8c1ac01a2e82410c9f4d59248c099/http.go#L969) in echoapp daprd logs I found this message:
`2021/01/14 16:12:50 error when serving connection "10.100.1.55:3500"<->"10.100.1.49:51830": cannot read multipart/form-data body: open /tmp/multipart-222081088: no such file or directory` it can't tell me much, but I think it is something related with permissions on daprd sidecar, or something related to fs.
I would appreciate any help, thank you.

**Comment by yaron2** [comment]
> Hi @yaron2
> I've made appropriate changes and during testing encountered on the two new issues.
> 
> 1. When I trying pass to my upload service via darp file more than 4 MB and less than 16 MB (I'll explain it in the next item) receiving the following error:
>    `{ "errorCode": "ERR_DIRECT_INVOKE", "message": "rpc error: code = ResourceExhausted desc = grpc: received message larger than max (10730238 vs. 4194304)" }`
>    Do you have any ideas on how can I fix that as well?
> 2. When I pass to my upload service via darp file more than 16 MB (seems that it is a default value for the maximum file size in memory https://github.com/valyala/fasthttp/blob/70e00dc4f3b8c1ac01a2e82410c9f4d59248c099/http.go#L969) in echoapp daprd logs I found this message:
>    `2021/01/14 16:12:50 error when serving connection "10.100.1.55:3500"<->"10.100.1.49:51830": cannot read multipart/form-data body: open /tmp/multipart-222081088: no such file or directory` it can't tell me much besides that fasthttp tries to save file more than 16 MB on tmp storage, but I think it is something related with permissions on daprd sidecar, or something related to fs.
>    I would appreciate any help, thank you.

Regarding 1, since Dapr uses gRPC to send messages in service invocation, the max size needs to propagate to the the code where Dapr starts its gRPC server and makes the gRPC call. In both cases it's a simple addition, you can see this issue that explains how to do it (the first comment shows the server side, and the second the client: https://github.com/grpc-ecosystem/grpc-gateway/issues/943)

The place where Dapr starts the gRPC server is here: https://github.com/dapr/dapr/blob/4678e477562e7e35a1d6ba04a9b17b6c9c00a025/pkg/grpc/server.go#L196

And here is the call to a remote Dapr sidecar that needs to include the client option: https://github.com/dapr/dapr/blob/944bcaeec10617021317e2123b129162689d19f9/pkg/messaging/direct_messaging.go#L160

**Comment by apuzyrevsky** [comment]
Hi, @yaron2 I figured out how to pass an argument to the dapr/pkg/grpc/server.go, and I did a hardcode for testing purposes in dapr/pkg/messaging/direct_messaging.go, it seems that resolved the issue,  but I am struggling to pass this property as a config argument. Is there any example that I can use as a reference? 

And if you have any thoughts regarding the second issue that I wrote, please share them.

**Comment by yaron2** [comment]
> Hi, @yaron2 I figured out how to pass an argument to the dapr/pkg/grpc/server.go, and I did a hardcode for testing purposes in dapr/pkg/messaging/direct_messaging.go, it seems that resolved the issue, but I am struggling to pass this property as a config argument. Is there any example that I can use as a reference?
> 
> And if you have any thoughts regarding the second issue that I wrote, please share them.

This is where you need to pass the new argument to the direct messaging package: https://github.com/dapr/dapr/blob/2849ede55102ff4c6a5da74de727b486f652f449/pkg/runtime/runtime.go#L456

Regarding the log error issue, it seems like the HTTP server is trying to write to a file and is denied because there's no filesystem mounted on the container. We (maintainers) will fix that in parallel to your PR.

**Comment by apuzyrevsky** [comment]
Hi, @yaron2 I made a PR in the WIP stage, feel free to comment on it, thank you.

**Comment by apuzyrevsky** [comment]
@yaron2 Could you, please notify me (for example it that issue) when you'll provide some changes from your side, so I can test the whole solution itself.

**Comment by ScottArbeit** [comment]
Just wanted to drop a thank you, @apuzyrevsky and @yaron2. I came here to search because I ran into this exact problem, and... thank you for adding this code. 🙏

**Comment by artursouza** [comment]
@apuzyrevsky @yaron2 Anything else pending in this Issue? Was it part of a previous release?

**Comment by apuzyrevsky** [comment]
> @apuzyrevsky @yaron2 Anything else pending in this Issue? Was it part of a previous release?

I am waiting for a reply from @yaron2 regarding "it seems like the HTTP server is trying to write to a file and is denied because there's no filesystem mounted on the container. We (maintainers) will fix that in parallel to your PR." if it is done or not.

**Comment by Betula-L** [comment]
@apuzyrevsky it seems that this issue has not been finished yet, while docs released.

**Comment by denis111** [comment]
Does Dapr buffer the whole request? If so than it's not right behaviour for uploading large files, it must act like a proxy then sending the stream of bytes to the destination service.
I'm also interested to be able to pass huge files without size restriction through Dapr.

**Comment by artursouza** [comment]
@Betula-L Please, where is the docs you are referring to?

@denis111 I agree that ideally we should stream requests and not buffer. Can you create an issue for that?

@apuzyrevsky Please, send a PR to add this configuration setting. Even if not ideal for a large file upload, it can unblock users that need this for a slightly larger payload than the default. This should be configurable for K8s too, not only standalone more, please.

Thanks.

**Comment by Betula-L** [comment]
https://docs.dapr.io/operations/configuration/increase-request-size/

**Comment by apuzyrevsky** [comment]
@artursouza will do that

**Comment by artursouza** [comment]
@Betula-L @apuzyrevsky  The problem is that the flag was added to daprd but not CLI. I am moving this issue to CLI repo.
